File: push/src/main/java/com/xiaomi/push/service/PushClientsManager.java
package com.xiaomi.push.service;

import android.content.Context;
import android.os.IBinder;
import android.os.Messenger;
import android.text.TextUtils;
import com.xiaomi.channel.commonutils.logger.MyLog;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;

/* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/push/service/PushClientsManager.class */
public class PushClientsManager {
    private static PushClientsManager sInstance;
    private final PushClientsCollection clients = new PushClientsCollection();

    /* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/push/service/PushClientsManager$ClientChangeListener.class */
    public interface ClientChangeListener {
        void onChange();
    }

    /* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/push/service/PushClientsManager$ClientLoginInfo.class */
    public static class ClientLoginInfo {
        public static final int TYPE_CHANNEL_CLOSE = 2;
        public static final int TYPE_CHANNEL_NO_NOTIFY = 0;
        public static final int TYPE_CHANNEL_OPEN_RESULT = 1;
        public static final int TYPE_CHANNEL_SERVER_KICK = 3;
        public String authMethod;
        public String chid;
        public String clientExtra;
        public String cloudExtra;
        public Context context;
        public boolean kick;
        public ClientEventDispatcher mClientEventDispatcher;
        private XMPushService mPushService;
        Messenger peer;
        public String pkgName;
        public String security;
        public String session;
        public String token;
        public String userId;
        ClientStatus status = ClientStatus.unbind;
        int currentRetrys = 0;
        final List<ClientStatusListener> statusChangeListeners = new ArrayList<>();
        ClientStatus notifiedStatus = null;
        boolean hasPeerSupport = false;
        final BindTimeoutJob timeOutJob = new BindTimeoutJob(this);
        IBinder.DeathRecipient peerWatcher = null;
        final PushClientNotifyJob notifyClientJob = new PushClientNotifyJob(this);

        /* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/push/service/PushClientsManager$ClientLoginInfo$ClientStatusListener.class */
        public interface ClientStatusListener {
            void onChange(ClientStatus clientStatus, ClientStatus clientStatus2, int i);
        }

        public ClientLoginInfo() {
        }

        public ClientLoginInfo(XMPushService xMPushService) {
            this.mPushService = xMPushService;
            addClientStatusListener(new ClientStatusListener() { // from class: com.xiaomi.push.service.PushClientsManager.ClientLoginInfo.1
                @Override // com.xiaomi.push.service.PushClientsManager.ClientLoginInfo.ClientStatusListener
                public void onChange(ClientStatus clientStatus, ClientStatus clientStatus2, int i) {
                    if (clientStatus2 == ClientStatus.binding) {
                        ClientLoginInfo.this.mPushService.executeJobDelayed(ClientLoginInfo.this.timeOutJob, 60000L);
                    } else {
                        ClientLoginInfo.this.mPushService.removeJobs(ClientLoginInfo.this.timeOutJob);
                    }
                }
            });
        }

        public static String getResource(String str) {
            if (TextUtils.isEmpty(str)) {
                return "";
            }
            int iLastIndexOf = str.lastIndexOf("/");
            return iLastIndexOf != -1 ? str.substring(iLastIndexOf + 1) : "";
        }

        XMPushService getPushService() {
            return this.mPushService;
        }

        void notifyClientStatus(int i, int i2, String str, String str2) {
            PushClientStatusSupport.notifyClientStatus(this, i, i2, str, str2);
        }

        boolean shouldNotifyClient(int i, int i2, String str) {
            return PushClientStatusSupport.shouldNotifyClient(this, i, i2, str);
        }

        public void addClientStatusListener(ClientStatusListener clientStatusListener) {
            synchronized (this.statusChangeListeners) {
                this.statusChangeListeners.add(clientStatusListener);
            }
        }

        public String getDesc(int i) {
            return PushClientStatusSupport.getDesc(i);
        }

        public long getNextRetryInterval() {
            return (((long) ((Math.random() * 20.0d) - 10.0d)) + ((long) ((this.currentRetrys + 1) * 15))) * 1000;
        }

        public void removeClientStatusListener(ClientStatusListener clientStatusListener) {
            synchronized (this.statusChangeListeners) {
                this.statusChangeListeners.remove(clientStatusListener);
            }
        }

        public void setStatus(ClientStatus clientStatus, int i, int i2, String str, String str2) {
            int i3;
            PushClientStatusSupport.notifyStatusListeners(this, clientStatus, i2);
            ClientStatus clientStatus2 = this.status;
            if (clientStatus2 != clientStatus) {
                MyLog.w(String.format("update the client %7$s status. %1$s->%2$s %3$s %4$s %5$s %6$s", clientStatus2, clientStatus, getDesc(i), PushConstants.getErrorDesc(i2), str, str2, this.chid));
                this.status = clientStatus;
            }
            if (this.mClientEventDispatcher == null) {
                MyLog.e("status changed while the client dispatcher is missing");
                return;
            }
            if (clientStatus == ClientStatus.binding) {
                return;
            }
            i3 = PushClientStatusSupport.computeNotifyDelay(this);
            this.mPushService.removeJobs(this.notifyClientJob);
            if (PushClientStatusSupport.isSpecialError(this, i, i2, str2)) {
                notifyClientStatus(i, i2, str, str2);
            } else {
                this.mPushService.executeJobDelayed(this.notifyClientJob.build(i, i2, str, str2), i3);
            }
        }

        void unwatch() {
            try {
                Messenger messenger = this.peer;
                if (messenger != null && this.peerWatcher != null) {
                    messenger.getBinder().unlinkToDeath(this.peerWatcher, 0);
                }
            } catch (Exception e) {
            }
            this.notifiedStatus = null;
        }

        void watch(Messenger messenger) {
            unwatch();
            try {
                if (messenger != null) {
                    this.peer = messenger;
                    this.hasPeerSupport = true;
                    this.peerWatcher = new PushClientPeerWatcher(this, messenger);
                    messenger.getBinder().linkToDeath(this.peerWatcher, 0);
                } else {
                    MyLog.i("peer linked with old sdk chid = " + this.chid);
                }
            } catch (Exception e) {
                MyLog.i("peer linkToDeath err: " + e.getMessage());
                this.peer = null;
                this.hasPeerSupport = false;
            }
        }

        void onPeerDied(final Messenger messenger) {
            this.mPushService.executeJobDelayed(new XMPushService.Job(0) {
                @Override
                public String getDesc() {
                    return "clear peer job";
                }

                @Override
                public void process() {
                    if (messenger == ClientLoginInfo.this.peer) {
                        MyLog.i("clean peer, chid = " + ClientLoginInfo.this.chid);
                        ClientLoginInfo.this.peer = null;
                    }
                }
            }, 0L);
            if ("9".equals(this.chid) && PushConstants.PUSH_SERVICE_PACKAGE_NAME.equals(this.mPushService.getPackageName())) {
                this.mPushService.executeJobDelayed(new XMPushService.Job(0) {
                    @Override
                    public String getDesc() {
                        return "check peer job";
                    }

                    @Override
                    public void process() {
                        PushClientsManager.ClientLoginInfo clientLoginInfoByChidAndUserId = PushClientsManager.getInstance().getClientLoginInfoByChidAndUserId(ClientLoginInfo.this.chid, ClientLoginInfo.this.userId);
                        if (clientLoginInfoByChidAndUserId != null && clientLoginInfoByChidAndUserId.peer == null) {
                            ClientLoginInfo.this.mPushService.closeChannel(ClientLoginInfo.this.chid, ClientLoginInfo.this.userId, 2, null, null);
                        }
                    }
                }, 60000L);
            }
        }
    }

    /* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/push/service/PushClientsManager$ClientStatus.class */
    public enum ClientStatus {
        unbind,
        binding,
        binded
    }

    private PushClientsManager() {
    }

    public static PushClientsManager getInstance() {
        PushClientsManager pushClientsManager;
        synchronized (PushClientsManager.class) {
            try {
                if (sInstance == null) {
                    sInstance = new PushClientsManager();
                }
                pushClientsManager = sInstance;
            } catch (Throwable th) {
                throw th;
            }
        }
        return pushClientsManager;
    }

    static String getSmtpLocalPart(String str) {
        if (TextUtils.isEmpty(str)) {
            return null;
        }
        int iIndexOf = str.indexOf("@");
        return iIndexOf > 0 ? str.substring(0, iIndexOf) : str;
    }

    public void addActiveClient(ClientLoginInfo clientLoginInfo) {
        synchronized (this) {
            this.clients.addActiveClient(clientLoginInfo);
        }
    }

    public void addClientChangeListener(ClientChangeListener clientChangeListener) {
        synchronized (this) {
            this.clients.addClientChangeListener(clientChangeListener);
        }
    }

    public void deactivateAllClientByChid(String str) {
        synchronized (this) {
            this.clients.deactivateAllClientByChid(str);
        }
    }

    public void deactivateClient(String str, String str2) {
        synchronized (this) {
            this.clients.deactivateClient(str, str2);
        }
    }

    public int getActiveClientCount() {
        int size;
        synchronized (this) {
            size = this.clients.getActiveClientCount();
        }
        return size;
    }

    public Collection<ClientLoginInfo> getAllClientLoginInfoByChid(String str) {
        synchronized (this) {
            return this.clients.getAllClientLoginInfoByChid(str);
        }
    }

    public ArrayList<ClientLoginInfo> getAllClients() {
        ArrayList<ClientLoginInfo> arrayList;
        synchronized (this) {
            arrayList = this.clients.getAllClients();
        }
        return arrayList;
    }

    public ClientLoginInfo getClientLoginInfoByChidAndUserId(String str, String str2) {
        synchronized (this) {
            return this.clients.getClientLoginInfoByChidAndUserId(str, str2);
        }
    }

    public void notifyConnectionFailed(Context context) {
        synchronized (this) {
            PushClientsStateSupport.notifyConnectionFailed(this.clients.getActiveClientMaps());
        }
    }

    public List<String> queryChannelIdByPackage(String str) {
        ArrayList<String> arrayList;
        synchronized (this) {
            arrayList = new ArrayList<>(PushClientsStateSupport.queryChannelIdByPackage(this.clients.getActiveClientMaps(), str));
        }
        return arrayList;
    }

    public void removeActiveClients() {
        synchronized (this) {
            this.clients.removeActiveClients();
        }
    }

    public void removeAllClientChangeListeners() {
        synchronized (this) {
            this.clients.removeAllClientChangeListeners();
        }
    }

    public void removeClientChangeListener(ClientChangeListener clientChangeListener) {
        synchronized (this) {
            this.clients.removeClientChangeListener(clientChangeListener);
        }
    }

    public void resetAllClients(Context context, int i) {
        synchronized (this) {
            PushClientsStateSupport.resetAllClients(this.clients.getActiveClientMaps(), i);
        }
    }
}

File: push/src/main/java/com/xiaomi/push/service/PushClientsManager.kt
package com.xiaomi.push.service

import android.content.Context
import android.os.IBinder
import android.os.Messenger
import android.text.TextUtils
import com.xiaomi.channel.commonutils.logger.MyLog
import java.util.ArrayList

class PushClientsManager private constructor() {
    private val clients = PushClientsCollection()

    fun interface ClientChangeListener {
        fun onChange()
    }

    class ClientLoginInfo() {
        @JvmField var authMethod: String = ""
        @JvmField var chid: String = ""
        @JvmField var clientExtra: String = ""
        @JvmField var cloudExtra: String = ""
        @JvmField var context: Context? = null
        @JvmField var kick: Boolean = false
        @JvmField var mClientEventDispatcher: ClientEventDispatcher? = null
        @JvmField var pkgName: String = ""
        @JvmField var security: String = ""
        @JvmField var session: String = ""
        @JvmField var token: String = ""
        @JvmField var userId: String = ""
        @JvmField var status: ClientStatus = ClientStatus.unbind
        @JvmField var currentRetrys: Int = 0
        @JvmField val statusChangeListeners: MutableList<ClientStatusListener> = ArrayList()
        @JvmField var notifiedStatus: ClientStatus? = null
        @JvmField var hasPeerSupport: Boolean = false
        @JvmField val timeOutJob: BindTimeoutJob = BindTimeoutJob(this)
        @JvmField var peerWatcher: IBinder.DeathRecipient? = null
        @JvmField val notifyClientJob: PushClientNotifyJob = PushClientNotifyJob(this)
        @JvmField var peer: Messenger? = null
        private var pushService: XMPushService? = null

        fun interface ClientStatusListener {
            fun onChange(previousStatus: ClientStatus, currentStatus: ClientStatus, reason: Int)
        }

        constructor(pushService: XMPushService) : this() {
            this.pushService = pushService
            addClientStatusListener { _, currentStatus, _ ->
                if (currentStatus == ClientStatus.binding) {
                    pushService.executeJobDelayed(timeOutJob, 60000L)
                } else {
                    pushService.removeJobs(timeOutJob)
                }
            }
        }

        fun getPushService(): XMPushService? = pushService

        fun notifyClientStatus(type: Int, reasonCode: Int, reasonMessage: String?, errorType: String?) {
            PushClientStatusSupport.notifyClientStatus(this, type, reasonCode, reasonMessage, errorType)
        }

        fun shouldNotifyClient(type: Int, reasonCode: Int, errorType: String?): Boolean {
            return PushClientStatusSupport.shouldNotifyClient(this, type, reasonCode, errorType)
        }

        fun addClientStatusListener(listener: ClientStatusListener) {
            synchronized(statusChangeListeners) {
                statusChangeListeners.add(listener)
            }
        }

        fun getDesc(type: Int): String {
            return PushClientStatusSupport.getDesc(type)
        }

        fun getNextRetryInterval(): Long {
            return (((Math.random() * 20.0) - 10.0).toLong() + ((currentRetrys + 1) * 15L)) * 1000L
        }

        fun removeClientStatusListener(listener: ClientStatusListener) {
            synchronized(statusChangeListeners) {
                statusChangeListeners.remove(listener)
            }
        }

        fun setStatus(clientStatus: ClientStatus, type: Int, reasonCode: Int, reasonMessage: String?, errorType: String?) {
            PushClientStatusSupport.notifyStatusListeners(this, clientStatus, reasonCode)
            val previousStatus = status
            if (previousStatus != clientStatus) {
                MyLog.w(
                    String.format(
                        "update the client %7\$s status. %1\$s->%2\$s %3\$s %4\$s %5\$s %6\$s",
                        previousStatus,
                        clientStatus,
                        getDesc(type),
                        PushConstants.getErrorDesc(reasonCode),
                        reasonMessage,
                        errorType,
                        chid,
                    ),
                )
                status = clientStatus
            }
            val dispatcher = mClientEventDispatcher
            if (dispatcher == null) {
                MyLog.e("status changed while the client dispatcher is missing")
                return
            }
            if (clientStatus == ClientStatus.binding) {
                return
            }
            val notifyDelay = PushClientStatusSupport.computeNotifyDelay(this)
            val service = pushService ?: return
            service.removeJobs(notifyClientJob)
            if (PushClientStatusSupport.isSpecialError(this, type, reasonCode, errorType)) {
                notifyClientStatus(type, reasonCode, reasonMessage, errorType)
            } else {
                service.executeJobDelayed(notifyClientJob.build(type, reasonCode, reasonMessage, errorType), notifyDelay.toLong())
            }
        }

        fun unwatch() {
            try {
                val currentPeer = peer
                val currentWatcher = peerWatcher
                if (currentPeer != null && currentWatcher != null) {
                    currentPeer.binder.unlinkToDeath(currentWatcher, 0)
                }
            } catch (_: Exception) {
            }
            notifiedStatus = null
        }

        fun watch(messenger: Messenger?) {
            unwatch()
            try {
                if (messenger != null) {
                    peer = messenger
                    hasPeerSupport = true
                    val watcher = PushClientPeerWatcher(this, messenger)
                    peerWatcher = watcher
                    messenger.binder.linkToDeath(watcher, 0)
                } else {
                    MyLog.i("peer linked with old sdk chid = $chid")
                }
            } catch (e: Exception) {
                MyLog.i("peer linkToDeath err: ${e.message}")
                peer = null
                hasPeerSupport = false
            }
        }

        fun onPeerDied(messenger: Messenger) {
            val service = pushService ?: return
            service.executeJobDelayed(
                object : XMPushService.Job(0) {
                    override fun getDesc(): String = "clear peer job"

                    override fun process() {
                        if (messenger == peer) {
                            MyLog.i("clean peer, chid = $chid")
                            peer = null
                        }
                    }
                },
                0L,
            )
            if (chid == "9" && PushConstants.PUSH_SERVICE_PACKAGE_NAME == service.packageName) {
                service.executeJobDelayed(
                    object : XMPushService.Job(0) {
                        override fun getDesc(): String = "check peer job"

                        override fun process() {
                            val current = getInstance().getClientLoginInfoByChidAndUserId(chid, userId)
                            if (current != null && current.peer == null) {
                                service.closeChannel(chid, userId, 2, null, null)
                            }
                        }
                    },
                    60000L,
                )
            }
        }

        companion object {
            const val TYPE_CHANNEL_NO_NOTIFY = 0
            const val TYPE_CHANNEL_OPEN_RESULT = 1
            const val TYPE_CHANNEL_CLOSE = 2
            const val TYPE_CHANNEL_SERVER_KICK = 3

            @JvmStatic
            fun getResource(userId: String?): String {
                if (TextUtils.isEmpty(userId)) {
                    return ""
                }
                val index = userId!!.lastIndexOf("/")
                return if (index != -1) userId.substring(index + 1) else userId
            }
        }
    }

    enum class ClientStatus {
        unbind,
        binding,
        binded,
    }

    fun addActiveClient(clientLoginInfo: ClientLoginInfo) {
        synchronized(this) {
            clients.addActiveClient(clientLoginInfo)
        }
    }

    fun addClientChangeListener(clientChangeListener: ClientChangeListener) {
        synchronized(this) {
            clients.addClientChangeListener(clientChangeListener)
        }
    }

    fun deactivateAllClientByChid(chid: String?) {
        if (chid == null) return
        synchronized(this) {
            clients.deactivateAllClientByChid(chid)
        }
    }

    fun deactivateClient(chid: String?, userId: String?) {
        if (chid == null || userId == null) return
        synchronized(this) {
            clients.deactivateClient(chid, userId)
        }
    }

    fun getActiveClientCount(): Int {
        synchronized(this) {
            return clients.getActiveClientCount()
        }
    }

    fun getAllClientLoginInfoByChid(chid: String?): Collection<ClientLoginInfo> {
        if (chid == null) return emptyList()
        synchronized(this) {
            return clients.getAllClientLoginInfoByChid(chid)
        }
    }

    fun getAllClients(): ArrayList<ClientLoginInfo> {
        synchronized(this) {
            return clients.getAllClients()
        }
    }

    fun getClientLoginInfoByChidAndUserId(chid: String?, userId: String?): ClientLoginInfo? {
        if (chid == null || userId == null) return null
        synchronized(this) {
            return clients.getClientLoginInfoByChidAndUserId(chid, userId)
        }
    }

    fun notifyConnectionFailed(context: Context) {
        synchronized(this) {
            PushClientsStateSupport.notifyConnectionFailed(clients.getActiveClientMaps())
        }
    }

    fun queryChannelIdByPackage(packageName: String?): List<String> {
        if (packageName == null) return emptyList()
        synchronized(this) {
            return ArrayList(PushClientsStateSupport.queryChannelIdByPackage(clients.getActiveClientMaps(), packageName))
        }
    }

    fun removeActiveClients() {
        synchronized(this) {
            clients.removeActiveClients()
        }
    }

    fun removeAllClientChangeListeners() {
        synchronized(this) {
            clients.removeAllClientChangeListeners()
        }
    }

    fun removeClientChangeListener(clientChangeListener: ClientChangeListener) {
        synchronized(this) {
            clients.removeClientChangeListener(clientChangeListener)
        }
    }

    fun resetAllClients(context: Context, reason: Int) {
        synchronized(this) {
            PushClientsStateSupport.resetAllClients(clients.getActiveClientMaps(), reason)
        }
    }

    companion object {
        @Volatile
        private var instance: PushClientsManager? = null

        @JvmStatic
        fun getInstance(): PushClientsManager {
            return instance ?: synchronized(PushClientsManager::class.java) {
                instance ?: PushClientsManager().also { instance = it }
            }
        }

        @JvmStatic
        fun getSmtpLocalPart(userId: String?): String? {
            if (TextUtils.isEmpty(userId)) {
                return null
            }
            val index = userId!!.indexOf("@")
            return if (index > 0) userId.substring(0, index) else userId
        }
    }
}
